File: crates/webhooks/examples/clients/consume_sse.py
"""
Subscribe to job events via SSE (Server-Sent Events).

pip install httpx-sse httpx
"""

import httpx
from httpx_sse import connect_sse
import json
import sys


def watch_job(base_url: str, job_id: str):
    """Stream job events until completion."""
    url = f"{base_url}/v1/jobs/{job_id}/events"

    with httpx.Client() as client:
        with connect_sse(client, "GET", url) as event_source:
            for event in event_source.iter_sse():
                data = json.loads(event.data)
                status = data["status"]
                print(f"[{event.event}] {json.dumps(data, indent=2)}")

                if status in ("completed", "failed", "cancelled"):
                    return data


def setup_webhook_receiver(port: int = 8080, secret: str = ""):
    """
    Minimal webhook receiver that verifies HMAC-SHA256 signatures.

    The operator POSTs to your URL with:
      X-Webhook-Signature: sha256=<hex>
      X-Job-Id: <job_id>
      Content-Type: application/json
    """
    from http.server import HTTPServer, BaseHTTPRequestHandler
    import hmac
    import hashlib

    class WebhookHandler(BaseHTTPRequestHandler):
        def do_POST(self):
            content_length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(content_length)

            # Verify HMAC signature
            sig_header = self.headers.get("X-Webhook-Signature", "")
            if secret and sig_header:
                expected_sig = sig_header.removeprefix("sha256=")
                computed = hmac.new(
                    secret.encode(), body, hashlib.sha256
                ).hexdigest()
                if not hmac.compare_digest(computed, expected_sig):
                    self.send_response(401)
                    self.end_headers()
                    self.wfile.write(b"invalid signature")
                    return

            job_id = self.headers.get("X-Job-Id", "unknown")
            payload = json.loads(body)
            print(f"[webhook] job={job_id} status={payload['event']['status']}")

            if payload["event"].get("result"):
                print(f"  result: {json.dumps(payload['event']['result'], indent=2)}")

            self.send_response(200)
            self.end_headers()

    print(f"Webhook receiver listening on :{port}")
    HTTPServer(("0.0.0.0", port), WebhookHandler).serve_forever()


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage:")
        print("  python consume_sse.py sse <base_url> <job_id>")
        print("  python consume_sse.py webhook <port> [secret]")
        sys.exit(1)

    mode = sys.argv[1]
    if mode == "sse":
        result = watch_job(sys.argv[2], sys.argv[3])
        print(f"\nFinal: {result['status']}")
    elif mode == "webhook":
        port = int(sys.argv[2]) if len(sys.argv) > 2 else 8080
        secret = sys.argv[3] if len(sys.argv) > 3 else ""
        setup_webhook_receiver(port, secret)

File: crates/webhooks/examples/clients/consume_sse.ts
/**
 * Subscribe to job events via SSE (Server-Sent Events).
 *
 * Works in Node.js 18+ (native fetch + EventSource) or browsers.
 * For Node.js < 18: npm install eventsource
 */

// ── SSE Consumer ────────────────────────────────────────────────────────

interface JobEvent {
  status: 'queued' | 'processing' | 'completed' | 'failed' | 'cancelled'
  progress?: number
  result?: Record<string, unknown>
  error?: string
  timestamp: number
}

/**
 * Watch a job via SSE until it reaches a terminal state.
 * Returns a promise that resolves with the final event.
 */
export function watchJob(baseUrl: string, jobId: string): Promise<JobEvent> {
  return new Promise((resolve, reject) => {
    const url = `${baseUrl}/v1/jobs/${jobId}/events`
    const source = new EventSource(url)

    const onEvent = (type: string) => (e: MessageEvent) => {
      const event: JobEvent = JSON.parse(e.data)
      console.log(`[${type}]`, JSON.stringify(event, null, 2))

      if (['completed', 'failed', 'cancelled'].includes(event.status)) {
        source.close()
        resolve(event)
      }
    }

    source.addEventListener('queued', onEvent('queued'))
    source.addEventListener('processing', onEvent('processing'))
    source.addEventListener('completed', onEvent('completed'))
    source.addEventListener('failed', onEvent('failed'))
    source.addEventListener('cancelled', onEvent('cancelled'))

    source.onerror = (e) => {
      source.close()
      reject(new Error(`SSE connection error: ${e}`))
    }
  })
}

// ── Webhook Receiver ────────────────────────────────────────────────────

import { createHmac, timingSafeEqual } from 'node:crypto'
import { createServer, type IncomingMessage, type ServerResponse } from 'node:http'

interface WebhookPayload {
  job_id: string
  event: JobEvent
}

/**
 * Verify HMAC-SHA256 signature from operator webhook delivery.
 */
function verifySignature(body: Buffer, signature: string, secret: string): boolean {
  const expected = signature.replace('sha256=', '')
  const computed = createHmac('sha256', secret).update(body).digest('hex')
  try {
    return timingSafeEqual(Buffer.from(computed, 'hex'), Buffer.from(expected, 'hex'))
  } catch {
    return false
  }
}

/**
 * Start a webhook receiver server.
 */
export function startWebhookReceiver(
  port: number,
  secret: string,
  onEvent: (payload: WebhookPayload) => void,
) {
  const server = createServer((req: IncomingMessage, res: ServerResponse) => {
    if (req.method !== 'POST') {
      res.writeHead(405)
      res.end()
      return
    }

    const chunks: Buffer[] = []
    req.on('data', (chunk: Buffer) => chunks.push(chunk))
    req.on('end', () => {
      const body = Buffer.concat(chunks)

      // Verify signature
      const sig = req.headers['x-webhook-signature'] as string
      if (secret && sig && !verifySignature(body, sig, secret)) {
        res.writeHead(401)
        res.end('invalid signature')
        return
      }

      const payload: WebhookPayload = JSON.parse(body.toString())
      onEvent(payload)

      res.writeHead(200)
      res.end('ok')
    })
  })

  server.listen(port, () => {
    console.log(`Webhook receiver listening on :${port}`)
  })

  return server
}

// ── CLI ─────────────────────────────────────────────────────────────────

const [,, mode, ...args] = process.argv

if (mode === 'sse') {
  const [baseUrl, jobId] = args
  watchJob(baseUrl, jobId).then((event) => {
    console.log(`\nFinal: ${event.status}`)
    process.exit(event.status === 'completed' ? 0 : 1)
  })
} else if (mode === 'webhook') {
  const port = parseInt(args[0] || '8080')
  const secret = args[1] || ''
  startWebhookReceiver(port, secret, (payload) => {
    console.log(`[webhook] job=${payload.job_id} status=${payload.event.status}`)
    if (payload.event.result) {
      console.log('  result:', JSON.stringify(payload.event.result, null, 2))
    }
  })
} else {
  console.log('Usage:')
  console.log('  npx tsx consume_sse.ts sse <base_url> <job_id>')
  console.log('  npx tsx consume_sse.ts webhook <port> [secret]')
}
